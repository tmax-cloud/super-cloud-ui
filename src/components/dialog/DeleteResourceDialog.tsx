/* eslint-disable no-useless-concat */
/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import * as React from 'react';
import { DialogContentProps } from './index';

import Button from '@mui/material/Button';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import DialogTitle from '@mui/material/DialogTitle';
import Alert from '@mui/material/Alert';
import ReportProblemIcon from '@mui/icons-material/ReportProblem';
import handleRequest from '../../apis/handleRequest';
import { RequestType } from '../../types';

export enum DialogSize {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

export type DeleteResourceDialogProps = DialogContentProps;

const alert = {
  error: css`
    .MuiPaper-root {
      width: 100%;
      height: auto;
    }
  `,
};

export default function DeleteResourceDialog(props: DeleteResourceDialogProps) {
  const { kindObj, setDialogOpen, resourceName, namespaceName } = props;

  const [error, setError] = React.useState<string>('');

  const deleteResourceMsg = (resourceName: string, namespaceName?: string) => {
    if (!!namespaceName) {
      return `Are you sure you want to delete ${resourceName} in namespace ${namespaceName}?`;
    } else {
      return `Are you sure you want to delete ${resourceName}?`;
    }
  };

  const onDeleteClick = async () => {
    try {
      await handleRequest(kindObj, RequestType.DELETE);
      onClose();
    } catch (e: any) {
      setError(e.message);
    }
  };

  const onClose = () => {
    setDialogOpen(false);
  };

  return (
    <>
      <DialogTitle>
        <ReportProblemIcon color="warning" fontSize="large" />
        Delete Resource Dialog
      </DialogTitle>
      <DialogContent>
        <DialogContentText>{deleteResourceMsg(resourceName, namespaceName)}</DialogContentText>
      </DialogContent>
      {error && (
        <div css={alert.error}>
          <Alert severity="error">{error}</Alert>
        </div>
      )}
      <DialogActions>
        <Button onClick={onClose} variant="outlined">
          Cancel
        </Button>
        <Button onClick={onDeleteClick} variant="contained" color="error">
          Confirm
        </Button>
      </DialogActions>
    </>
  );
}
