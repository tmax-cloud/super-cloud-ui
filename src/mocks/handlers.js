import { rest } from 'msw';

export const handlers = [
  rest.get('/api/kubernetes/api/v1/services/', (req, res, ctx) => {
    return res(
      ctx.json({
        kind: 'ServiceList',
        apiVersion: 'v1',
        metadata: {
          selfLink: '/api/v1/services',
          resourceVersion: '105697845',
        },
        items: [
          {
            metadata: {
              name: 'api-gateway',
              namespace: 'api-gateway-system',
              selfLink: '/api/v1/namespaces/api-gateway-system/services/api-gateway',
              uid: '3169d866-b4a8-4b24-8064-3e2766cf9919',
              resourceVersion: '1211332',
              creationTimestamp: '2021-12-13T02:40:18Z',
              labels: {
                'app.kubernetes.io/component': 'reverse-proxy',
                'app.kubernetes.io/managed-by': 'tmax-cloud',
                'app.kubernetes.io/name': 'traefik',
                'app.kubernetes.io/part-of': 'api-gateway',
              },
              annotations: {
                'kubectl.kubernetes.io/last-applied-configuration':
                  '{"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"app.kubernetes.io/component":"reverse-proxy","app.kubernetes.io/managed-by":"tmax-cloud","app.kubernetes.io/name":"traefik","app.kubernetes.io/part-of":"api-gateway"},"name":"api-gateway","namespace":"api-gateway-system"},"spec":{"ports":[{"name":"traefik","port":9000,"protocol":"TCP","targetPort":"traefik"},{"name":"web","port":80,"protocol":"TCP","targetPort":"web"},{"name":"websecure","port":443,"protocol":"TCP","targetPort":"websecure"},{"name":"k8s","port":6443,"protocol":"TCP","targetPort":"k8s"}],"selector":{"app":"traefik","app.kubernetes.io/component":"reverse-proxy","app.kubernetes.io/managed-by":"tmax-cloud","app.kubernetes.io/name":"traefik","app.kubernetes.io/part-of":"api-gateway"},"sessionAffinity":"None","type":"LoadBalancer"}}\n',
              },
              managedFields: [
                {
                  manager: 'controller',
                  operation: 'Update',
                  apiVersion: 'v1',
                  time: '2021-12-13T02:40:18Z',
                  fieldsType: 'FieldsV1',
                  fieldsV1: {
                    'f:status': {
                      'f:loadBalancer': {
                        'f:ingress': {},
                      },
                    },
                  },
                },
                {
                  manager: 'kubectl-client-side-apply',
                  operation: 'Update',
                  apiVersion: 'v1',
                  time: '2021-12-13T02:40:18Z',
                  fieldsType: 'FieldsV1',
                  fieldsV1: {
                    'f:metadata': {
                      'f:annotations': {
                        '.': {},
                        'f:kubectl.kubernetes.io/last-applied-configuration': {},
                      },
                      'f:labels': {
                        '.': {},
                        'f:app.kubernetes.io/component': {},
                        'f:app.kubernetes.io/managed-by': {},
                        'f:app.kubernetes.io/name': {},
                        'f:app.kubernetes.io/part-of': {},
                      },
                    },
                    'f:spec': {
                      'f:externalTrafficPolicy': {},
                      'f:ports': {
                        '.': {},
                        'k:{"port":80,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                        'k:{"port":443,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                        'k:{"port":6443,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                        'k:{"port":9000,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                      },
                      'f:selector': {
                        '.': {},
                        'f:app': {},
                        'f:app.kubernetes.io/component': {},
                        'f:app.kubernetes.io/managed-by': {},
                        'f:app.kubernetes.io/name': {},
                        'f:app.kubernetes.io/part-of': {},
                      },
                      'f:sessionAffinity': {},
                      'f:type': {},
                    },
                  },
                },
              ],
            },
            spec: {
              ports: [
                {
                  name: 'traefik',
                  protocol: 'TCP',
                  port: 9000,
                  targetPort: 'traefik',
                  nodePort: 31751,
                },
                {
                  name: 'web',
                  protocol: 'TCP',
                  port: 80,
                  targetPort: 'web',
                  nodePort: 30958,
                },
                {
                  name: 'websecure',
                  protocol: 'TCP',
                  port: 443,
                  targetPort: 'websecure',
                  nodePort: 30860,
                },
                {
                  name: 'k8s',
                  protocol: 'TCP',
                  port: 6443,
                  targetPort: 'k8s',
                  nodePort: 31407,
                },
              ],
              selector: {
                app: 'traefik',
                'app.kubernetes.io/component': 'reverse-proxy',
                'app.kubernetes.io/managed-by': 'tmax-cloud',
                'app.kubernetes.io/name': 'traefik',
                'app.kubernetes.io/part-of': 'api-gateway',
              },
              clusterIP: '10.96.225.224',
              type: 'LoadBalancer',
              sessionAffinity: 'None',
              externalTrafficPolicy: 'Cluster',
            },
            status: {
              loadBalancer: {
                ingress: [
                  {
                    ip: '192.168.9.141',
                  },
                ],
              },
            },
          },
          {
            metadata: {
              name: 'console',
              namespace: 'api-gateway-system',
              selfLink: '/api/v1/namespaces/api-gateway-system/services/console',
              uid: '9a1f0543-ed1a-4335-a604-6ca898dada0c',
              resourceVersion: '96329224',
              creationTimestamp: '2022-02-09T23:26:29Z',
              labels: {
                'app.kubernetes.io/component': 'webserver',
                'app.kubernetes.io/instance': 'console-ck-master',
                'app.kubernetes.io/managed-by': 'Helm',
                'app.kubernetes.io/name': 'console',
                'app.kubernetes.io/part-of': 'api-gateway',
                'app.kubernetes.io/version': '5.0.42.4',
                'helm.sh/chart': 'console-0.1.0',
              },
              annotations: {
                createdTime: '2022-02-10T08:26:29.288982552+09:00',
                creator: 'kubernetes-admin',
                'kubectl.kubernetes.io/last-applied-configuration':
                  '{"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"app.kubernetes.io/component":"webserver","app.kubernetes.io/instance":"console-ck-master","app.kubernetes.io/managed-by":"Helm","app.kubernetes.io/name":"console","app.kubernetes.io/part-of":"api-gateway","app.kubernetes.io/version":"5.0.42.4","helm.sh/chart":"console-0.1.0"},"name":"console","namespace":"api-gateway-system"},"spec":{"ports":[{"name":"http","port":31303,"protocol":"TCP","targetPort":"http"}],"selector":{"app.kubernetes.io/instance":"console-ck-master","app.kubernetes.io/name":"console"},"type":"ClusterIP"}}\n',
                updatedTime: '2022-02-10T08:26:29.288982552+09:00',
                updater: 'kubernetes-admin',
              },
              managedFields: [
                {
                  manager: 'argocd-application-controller',
                  operation: 'Update',
                  apiVersion: 'v1',
                  time: '2022-02-09T23:26:29Z',
                  fieldsType: 'FieldsV1',
                  fieldsV1: {
                    'f:metadata': {
                      'f:annotations': {
                        '.': {},
                        'f:kubectl.kubernetes.io/last-applied-configuration': {},
                      },
                      'f:labels': {
                        '.': {},
                        'f:app.kubernetes.io/component': {},
                        'f:app.kubernetes.io/instance': {},
                        'f:app.kubernetes.io/managed-by': {},
                        'f:app.kubernetes.io/name': {},
                        'f:app.kubernetes.io/part-of': {},
                        'f:app.kubernetes.io/version': {},
                        'f:helm.sh/chart': {},
                      },
                    },
                    'f:spec': {
                      'f:ports': {
                        '.': {},
                        'k:{"port":31303,"protocol":"TCP"}': {
                          '.': {},
                          'f:name': {},
                          'f:port': {},
                          'f:protocol': {},
                          'f:targetPort': {},
                        },
                      },
                      'f:selector': {
                        '.': {},
                        'f:app.kubernetes.io/instance': {},
                        'f:app.kubernetes.io/name': {},
                      },
                      'f:sessionAffinity': {},
                      'f:type': {},
                    },
                  },
                },
              ],
            },
            spec: {
              ports: [
                {
                  name: 'http',
                  protocol: 'TCP',
                  port: 31303,
                  targetPort: 'http',
                },
              ],
              selector: {
                'app.kubernetes.io/instance': 'console-ck-master',
                'app.kubernetes.io/name': 'console',
              },
              clusterIP: '10.96.229.111',
              type: 'ClusterIP',
              sessionAffinity: 'None',
            },
            status: {
              loadBalancer: {},
            },
          },
        ],
      }),
    );
  }),
];
