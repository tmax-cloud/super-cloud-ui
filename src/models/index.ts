export interface K8sModelType {
  apiVersion: string;
  label: string;
  plural: string;
  abbr: string;
  namespaced: boolean;
  kind: string;
  id: string;
  labelPlural: string;
}

export const ServiceModel: K8sModelType = {
  apiVersion: 'v1',
  label: 'Service',
  plural: 'services',
  abbr: 'S',
  namespaced: true,
  kind: 'Service',
  id: 'service',
  labelPlural: 'Services',
};
