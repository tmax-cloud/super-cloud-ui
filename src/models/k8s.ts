import { K8sKind } from '../types';

export const ServiceModel: K8sKind = {
  apiVersion: 'v1',
  label: 'Service',
  plural: 'services',
  abbr: 'S',
  namespaced: true,
  kind: 'Service',
  id: 'service',
  labelPlural: 'Services',
};
