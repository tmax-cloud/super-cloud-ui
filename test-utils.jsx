// import React from 'react';
// import { render } from '@testing-library/react';
// 추후에 redux 추가하게 되면 redux Provider로 감싸주는 custom render 메서드 추가 해야됨.

export * from '@testing-library/react';
