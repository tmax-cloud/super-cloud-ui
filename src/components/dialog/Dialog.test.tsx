import { rest } from 'msw';
import userEvent from '@testing-library/user-event';
import { render, screen } from '../../../test-utils';
import { server } from '../../mocks/server';
import { DialogSize, DialogOpenButton } from './index';
import DeleteResourceDialog from './DeleteResourceDialog';
import { ServiceModel } from '../../models/index';

function setup() {
  return render(<DialogOpenButton SubComponent={DeleteResourceDialog} subProps={{ resourceName: 'podName_01', namespaceName: 'namespaceName_01', kindObj: { ServiceModel } }} size={DialogSize.medium} />);
}

describe('삭제 다이알로그 테스트', () => {
  test('테이블 스냅샷 테스트', async () => {
    setup();
    const openDialogBtn = screen.getByRole('button');
    userEvent.click(openDialogBtn);
    const dialog = await screen.findByRole('dialog');
    expect(dialog).toMatchSnapshot();
  });

  test('delete Dialog에서 에러 발생시 validation 문구 테스트', async () => {
    server.use(
      rest.delete('/api/kubernetes/api/v1/namespaces/test_ns/services/test_resource', (req, res, ctx) => {
        return res.once(ctx.status(404), ctx.json({ errorMessage: 'server Error' }));
      }),
    );
    setup();
    const openDialogBtn = screen.getByRole('button');
    userEvent.click(openDialogBtn);

    const confirmButton = screen.getByText('Confirm');
    userEvent.click(confirmButton);

    const alert = await screen.findByRole('alert');
    expect(alert).toBeInTheDocument();
  });
});
