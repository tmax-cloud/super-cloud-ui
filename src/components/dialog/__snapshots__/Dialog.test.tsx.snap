// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`삭제 다이알로그 테스트 테이블 스냅샷 테스트 1`] = `
<div
  aria-labelledby="mui-1"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
  role="dialog"
>
  <h2
    class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
    id="mui-1"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-colorWarning MuiSvgIcon-fontSizeLarge css-1uo3q5b-MuiSvgIcon-root"
      data-testid="ReportProblemIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
      />
    </svg>
    Delete Resource Dialog
  </h2>
  <div
    class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
  >
    <p
      class="MuiDialogContentText-root MuiTypography-root MuiTypography-body1 css-qfso29-MuiTypography-root-MuiDialogContentText-root"
    >
      Are you sure you want to delete undefined?
    </p>
  </div>
  
  <div
    class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
  >
    <button
      class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Cancel
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-1sszw6i-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Confirm
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
